saleor/graphql/order/mutations/fulfillments.py

```diff
@@ -643,21 +643,105 @@ def clean_input(cls, info, order_id, input):
+    def remove_waiting_fulfillments(cls, lines_info):
+        """Delete fulfillment lines which fulfillments are `WAITING_FOR_APPROVAL`.
+
+        Delete fulfillments themselves if metioned lines were only lines inside.
+        """
+        if not lines_info:
+            return {}, []
+
+        lines_to_remove = []
+        lines_to_update = []
+        remaining_fulfillment_lines = []
+        order_lines_map = defaultdict(lambda: 0)
+
+        for line_data in lines_info:
+            if (
+                line_data.line.fulfillment.status
+                == FulfillmentStatus.WAITING_FOR_APPROVAL
+            ):
+                if line_data.quantity < line_data.line.quantity:
+                    _line = line_data.line
+                    _line.quantity = _line.quantity - line_data.quantity
+                    lines_to_update.append(_line)
+                else:
+                    lines_to_remove.append(line_data.line.id)
+                order_lines_map[line_data.line.order_line] += line_data.quantity
+            else:
+                remaining_fulfillment_lines.append(line_data)
+
+        lines_to_delete = order_models.FulfillmentLine.objects.filter(
+            id__in=set(lines_to_remove)
+        ).select_related("fulfillment")
+
+        related_fulfillment_ids = [f.fulfillment.pk for f in lines_to_delete]
+        related_fulfillment_ids.extend([f.fulfillment.pk for f in lines_to_update])
+
+        if lines_to_update:
+            order_models.FulfillmentLine.objects.bulk_update(
+                lines_to_update, ["quantity"]
+            )
+
+        if lines_to_delete:
+            lines_to_delete.delete()
+
+        order_models.Fulfillment.objects.filter(
+            pk__in=related_fulfillment_ids, lines=None
+        ).delete()
+
+        return order_lines_map, remaining_fulfillment_lines
+
+    @classmethod
+    def extend_order_lines(cls, order_lines, lines_data):
+        """Extend order_lines with data from lines_data.
+
+        If order line entry existed, increase it's quantity, create if it didn't.
+        """
+        for line, quantity in lines_data.items():
+            appended = False
+            for item in order_lines:
+                if item.line == line:
+                    item.quantity += quantity
+                    appended = True
+            if not appended:
+                order_lines.append(
+                    OrderLineData(
+                        line=line,
+                        quantity=quantity,
+                        variant=None,
+                        replace=False,
+                        warehouse_pk=None,
+                    )
+                )
+        return order_lines
+
     @classmethod
     def perform_mutation(cls, _root, info, **data):
         cleaned_input = cls.clean_input(info, data.get("order"), data.get("input"))
         order = cleaned_input["order"]
+
+        order_lines_map, remaining_fulfillment_lines = cls.remove_waiting_fulfillments(
```

Comment:
Not sure of this (or maybe I am missing something).  Why do we have to do that - `cls.remove_waiting_fulfillments` and `cls.extend_order_lines`?
Wouldn't be good to treat fulfillment `Waiting for Approving` as a fulfillment without allocated stock, and extend a logic inside `create_refund_fulfillment` to properly handle it (especially `saleor.order.actions._move_fulfillment_lines_to_target_fulfillment`)?  
It should work by default.

Reply:
That was my frist approach but there is already way more complicated logic there, which also touches stocks and stuff based on quantity allocated (which doesn't include waiting ones), it was hard to get it to even work and calculate things properly, plus it made a mess with extra `if`s in already complicated places, thats why it's now more "chained".

Reply:
I understand that we don't want to allocate lines from `waiting for approving`, but is it OK that we don't update quantity_fulfilled after we move lines to the fulfillment - `waiting for approving`?
Personally, I would try to extend `_move_fulfillment_lines_to_target_fulfillment` even if this will include some ifs or complicates the logic.  This should also solve the issue with return/replace case. In your case `return/replace` will also require such additional actions. For me, it looks like we change the data in order and in fulfillments to fit in old logic for refunds. 
Also, the assumption of `create_refund_fulfillment` is that you can call it and it will cover for you all things related to refund, but it will not work well with the new fulfillment type.